Document join reordering by the optimizer

rmloveland opened the pull request:
Fixes #4330.

Summary of changes:

- Update 'Cost-based optimizer' page with a new section 'Join
  reordering', which discusses the new functionality, as well as
  tradeoffs involved in its operation.  The new
  `experimental_reorder_joins_limit` setting is described, along with
  some warnings to the user about not setting it too high, since that
  may lead to performance degradation.

- Update the session variables page to include the
  `experimental_reorder_joins_limit` variable.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/4332)
<!-- Reviewable:end -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/98fffc974965e241c6d6f240e27c1e1cb3ec8461/

rmloveland commented:
For convenience, direct link to the new [join reordering](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/98fffc974965e241c6d6f240e27c1e1cb3ec8461/dev/cost-based-optimizer.html#join-reordering) section of the CBO page.

Specific questions:

- Is the mention of "join associativity" too specific?  I read the PR which maybe I shouldn't have since I didn't "understand" it
- Is it accurate to say the increase in the number of possible plans is exponential, as we do in this draft?
- I made up a max number for the setting (12), but what should the max recommended value be?
- Is there anything incorrect?
- Is there anything missing that you think should be stated here?

justinj commented on v2.2/cost-based-optimizer.md:
```diff
@@ -143,6 +143,25 @@ The query plan cache is still under development and has the following limitation
+We strongly recommend not setting this value higher than 12 to avoid performance degradation. If set too high, the cost of generating and costing execution plans can end up dominating the total execution time of the query.
+{{site.data.alerts.end}}
+
+For more information about the difficulty of selecting an optimal join ordering (spoiler: it's NP-complete), see our blog post [An Introduction to Join Ordering](https://www.cockroachlabs.com/blog/join-ordering-pt1/).
```
I'm not actually sure it's NP-complete, it's definitely NP-hard.

justinj commented on v2.2/cost-based-optimizer.md:
```diff
@@ -143,6 +143,25 @@ The query plan cache is still under development and has the following limitation
+
+<span class="version-tag">New in v2.2</span>: The cost-based optimizer will explore additional join orderings in an attempt to find the lowest-cost execution plan for a query involving multiple joins. Specifically, it applies the rule of join associativity to generate even more possible join orderings to choose from, which can lead to significantly better performance in some cases.
+
+Because this process leads to an exponential increase in the number of possible execution plans for such queries, it's only used on queries with 4 or fewer joins by default, as determined by the `experimental_reorder_joins_limit` [session variable](set-vars.html).
```
It's not really "queries with 4 or fewer joins", it's "will reorder subtrees containing 4 or fewer joins." But I'm not sure if that's difficult to communicate here?

The difference is that if a query has, say, 8 joins, the subtree it has containing 4 *will* be reordered.

rmloveland commented:
> On Feb 6, 2019, at 5:24 PM, Justin Jeffrey wrote:
> 
> @justinj approved this pull request.
> 
> I'm biased towards being less specific, but if you want to include the word "associativity" you should also include "commutativity" since they work together to explore everything.
I removed the mention of "associativity" then, since having both of these terms means we should probably get into what they mean, which is really handled better in your blog post(s), for now.  It would def blow up our word count here and be possibly confusing for non-join-experts.

> It's actually worse, it follows the Catalan Numbers <https://en.wikipedia.org/wiki/Catalan_number>. But exponential is close enough for these purposes I think.
OK, leaving it at “exponential” then. 🙂
> 12 is super high, I would say going above 8 will begin to cause problems. Here's a rough graph I made showing the plan time for max depth+number of tables a couple weeks ago:
> 
OK, will update to say the limit should be 8.

Related: should we add a line that says “Most users should not change this setting” ?   It might be good to discourage casually tweaking this. (Or not? What do you think?)
>  <https://user-images.githubusercontent.com/409075/52377493-0c535f80-2a33-11e9-92f1-d951622e5bff.png>
> In v2.2/cost-based-optimizer.md <https://github.com/cockroachdb/docs/pull/4332#discussion_r254475852>:
> 
> > +<span class="version-tag">New in v2.2</span>: The cost-based optimizer will explore additional join orderings in an attempt to find the lowest-cost execution plan for a query involving multiple joins. Specifically, it applies the rule of join associativity to generate even more possible join orderings to choose from, which can lead to significantly better performance in some cases.
> +
> +Because this process leads to an exponential increase in the number of possible execution plans for such queries, it's only used on queries with 4 or fewer joins by default, as determined by the `experimental_reorder_joins_limit` [session variable](set-vars.html).
> +
> +To change the number of joins that will be reordered by the optimizer, set the variable as shown below.
> +
> +{% include copy-clipboard.html %}
> +~~~ sql
> +> SET experimental_reorder_joins_limit = 8;
> +~~~
> +
> +{{site.data.alerts.callout_danger}}
> +We strongly recommend not setting this value higher than 12 to avoid performance degradation. If set too high, the cost of generating and costing execution plans can end up dominating the total execution time of the query.
> +{{site.data.alerts.end}}
> +
> +For more information about the difficulty of selecting an optimal join ordering (spoiler: it's NP-complete), see our blog post [An Introduction to Join Ordering](https://www.cockroachlabs.com/blog/join-ordering-pt1/).
> I'm not actually sure it's NP-complete, it's definitely NP-hard.
> 
Ah, typo on my part since I didn’t really know the difference!  Updated to say “NP-hard”, and made it a link to Wikipedia article on that topic for the curious (which I will now go read 🙂)

> 
> In v2.2/cost-based-optimizer.md <https://github.com/cockroachdb/docs/pull/4332#discussion_r254476689>:
> 
> > @@ -143,6 +143,25 @@ The query plan cache is still under development and has the following limitation
>  - If you use the query plan cache in conjunction with table statistics, cached plans do not yet get invalidated when new statistics are created.
>  {{site.data.alerts.end}}
>  
> +## Join reordering
> +
> +<span class="version-tag">New in v2.2</span>: The cost-based optimizer will explore additional join orderings in an attempt to find the lowest-cost execution plan for a query involving multiple joins. Specifically, it applies the rule of join associativity to generate even more possible join orderings to choose from, which can lead to significantly better performance in some cases.
> +
> +Because this process leads to an exponential increase in the number of possible execution plans for such queries, it's only used on queries with 4 or fewer joins by default, as determined by the `experimental_reorder_joins_limit` [session variable](set-vars.html).
> It's not really "queries with 4 or fewer joins", it's "will reorder subtrees containing 4 or fewer joins." But I'm not sure if that's difficult to communicate here?
> 

If it’s what this setting refers to, I think we must explain it - what does “subtree” mean in this context from a user’s POV?  Do I need to look at EXPLAIN output, or is it possible to determine by visually inspecting a SQL query?
> The difference is that if a query has, say, 8 joins, the subtree it has containing 4 will be reordered.
> 
I can definitely add this info once we explain subtrees (as discussed above).

Finally, I’ve pushed the changes I’ve made that incorporate your feedback so far.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/524c67d7608f47956b3fdfec0584a1efc8e9b446/

rmloveland commented:
Re: discussion above about describing "subtrees", I think that work can wait until after the current alpha release.  Leaving this as is for now for docs review.

rmloveland commented:
> I'd remove `(spoiler: it's NP-hard)`.

Fixed.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/a98524617d84ef727646d7aac22264e81cc9b5b4/
